exec/cnex/lib/sqlite.c

```diff
@@ -0,0 +1,382 @@
+    r->ptr = db;
+    r->release = object_releaseDatabaseObject;
+    r->toString = object_databaseObjectToString;
+
+    return r;
+}
+
+
+
+typedef struct tagTCursor {
+    TExecutor *exec;
+    sqlite3 *db;
+    char *query;
+    sqlite3_stmt *stmt;
+    int columns;
+} Cursor;
+
+static void openCursor(Cursor *self)
+{
+    int r = sqlite3_prepare_v2(self->db, self->query, -1, &self->stmt, NULL);
+    if (r != SQLITE_OK) {
+        self->exec->rtl_raise(self->exec , "SqlException", sqlite3_errmsg(self->db), number_from_sint32(r));
+        return;
+    }
+    self->columns = sqlite3_column_count(self->stmt);
+}
+
+static BOOL fetchCursor(Cursor *self, Cell *result)
+{
+    int r = sqlite3_step(self->stmt);
+    if (r == SQLITE_DONE) {
+        return FALSE;
+    }
+    if (r == SQLITE_ROW) {
+        for (int i = 0; i < self->columns; i++) {
+            Cell *col = cell_fromCString((const char*)sqlite3_column_text(self->stmt, i));
+            cell_arrayAppendElementPointer(result, col);
+        }
+    } else {
+        self->exec->rtl_raise(self->exec, "SqlException", sqlite3_errmsg(self->db), number_from_sint32(r));
+        return FALSE;
+    }
+    return TRUE;
+}
+
+static void closeCursor(Cursor *self)
+{
+    sqlite3_finalize(self->stmt);
+    self->stmt = NULL;
+}
+
+static void freeCursor(Cursor *self)
+{
+    if (self->stmt) {
+        closeCursor(self);
+    }
+    free(self->query);
+    free(self);
+}
+
+static Cursor *declareCursor(TExecutor *exec, sqlite3 *db, char *query)
+{
+    Cursor *r = malloc(sizeof(Cursor));
+    if (r == NULL) {
+        fatal_error("Could not alloc SQLite Cursor.\n");
+    }
+
+    r->db = db;
+    r->query = query;
+    r->stmt = NULL;
+    r->columns = 0;
+    r->exec = exec;
+
+    return r;
+}
+
+static int callback(void *rowscell, int columns, char **values, char **names)
+{
+    (void)names; // prevent compiler warning "expected formal parameter list, not a type list"
+    Cell *row = cell_createArrayCell(columns);
+    for (int i = 0; i < columns; i++) {
+        row->array->data[i].type = cString;
+        row->array->data[i].string = string_createCString(values[i]);
+    }
+    cell_arrayAppendElementPointer(rowscell, row);
+    return 0;
+}
+
+
+
+void sqlite_open(TExecutor *exec)
+{
+    char *name = string_asCString(top(exec->stack)->string); pop(exec->stack);
+
+    sqlite3 *db;
+    int r = sqlite3_open(name, &db);
+    if (r != SQLITE_OK) {
+        exec->rtl_raise(exec, "SqlException", sqlite3_errmsg(db), number_from_sint64(r));
+        free(name);
+        return;
+    }
+
+    push(exec->stack, cell_fromObject(object_createDatabaseObject(db)));
+    free(name);
+}
+
+void sqlite_exec(TExecutor *exec)
+{
+    Cell *parameters = cell_fromCell(top(exec->stack)); pop(exec->stack);
+    char *sql = string_asCString(top(exec->stack)->string); pop(exec->stack);
+    Cell *pdb = cell_fromCell(top(exec->stack)); pop(exec->stack);
+
+    Object *db = check_database(exec, pdb->object);
```

Comment:
Shouldn't you check the return value of this function here?

Reply:
Changed all references to check if db is null, and return from the function right away.